dataplane/protocol/lldp/lldp.go

```diff
@@ -0,0 +1,293 @@
+	return IsLldp(po)
+}
+
+// Process dispatches the packet to the corresponding port handler.
+func (d *Daemon) Process(p *packetio.Packet) error {
+	pd, ok := d.portDaemons[fmt.Sprintf("%d", p.HostPort)]
+	if !ok {
+		return fmt.Errorf("port %q not found", p.HostPort)
+	}
+	return pd.Process(p)
+}
+
+// portDaemon contains the required information for LLDP and processes the LLDP frames for a given hostif.
+type portDaemon struct {
+	Name      string
+	info      *lldpInfo
+	Interval  time.Duration
+	doneCh    chan struct{}
+	inCh      chan *packetio.Packet
+	errRecvCh chan error
+}
+
+// newPortDaemon creates a port daemon to send/recv LLDP protocol.
+func newPortDaemon(hostif string) *portDaemon {
+	d := &portDaemon{
+		Name:      hostif,
+		doneCh:    make(chan struct{}),
+		inCh:      make(chan *packetio.Packet),
+		errRecvCh: make(chan error),
+	}
+	go func() {
+		log.Infof("Start LLDP sender.")
+		for {
+			select {
+			case <-d.doneCh:
+				log.Infof("Stop sending LLDP frame")
+				return
+			default:
+				_, err := d.info.Frame()
+				if err != nil {
+					log.Errorf("failed to create LLDP frame: %v", err)
+					continue
+				}
+				// TODO: Send the packe to the hostif.
+				log.Infof("Write LLDP frame to port: %+v", d.Name)
+				time.Sleep(d.Interval)
+			}
+		}
+	}()
+
+	go func() {
+		log.Infof("Start LLDP receiver.")
+		for {
+			select {
+			case <-d.doneCh:
+				log.Infof("Stop processing LLDP frame")
+				return
+			default:
+				f := <-d.inCh
+				log.Infof("Got LLDP Frame: %+v", f.String())
+				pkt := gopacket.NewPacket(f.GetFrame(), layers.LayerTypeEthernet, gopacket.Default)
+				for _, layer := range pkt.Layers() {
+					if layer.LayerType() != layers.LayerTypeLinkLayerDiscoveryInfo {
+						log.Infof("Skipped layer %v", layer.LayerType().String())
+						continue
+					}
+					info, ok := layer.(*layers.LinkLayerDiscoveryInfo)
+					if !ok {
+						d.errRecvCh <- fmt.Errorf("packet is not LinkLayerDiscoveryInfo: %+v", layer)
+						continue
+					}
+					d.info.RemoteSysName = info.SysName
+					d.info.RemoteSysDesc = info.SysDescription
+				}
+			}
+		}
+	}()
+	return d
+}
+
+// Process handles the packet from the hostif and update the remote information.
+func (d *portDaemon) Process(p *packetio.Packet) error {
+	if d.inCh == nil {
+		return fmt.Errorf("failed to inject packet")
+	}
+	d.inCh <- p
+	err := <-d.errRecvCh
+	return err
+}
+
+// Stop stops the port daemon.
+func (d *portDaemon) Stop() {
+	d.doneCh <- struct{}{}
+}
+
+// IsLldp returns whether the packet is a LLDP frame.
+func IsLldp(po *packetio.PacketOut) bool {
+	ethPkt := gopacket.NewPacket(po.GetPacket().GetFrame(), layers.LayerTypeEthernet, gopacket.Default)
+	ethLayer := ethPkt.Layer(layers.LayerTypeEthernet)
+	if ethLayer == nil {
+		return false
+	}
+	return ethLayer.(*layers.Ethernet).EthernetType == layers.EthernetType(layers.EthernetTypeLinkLayerDiscovery.LayerType())
+}
+
+// lldpInfo contains LLDP protocol information.
+type lldpInfo struct {
```

Comment:
no action needed now, but in a future PR, we could maybe use the OC  type directly here

Reply:
Ack